### PR TITLE
[extractor/crackle] Adds support for channels, playlists, and series

### PR DESCRIPTION
**Pull request by shloop**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Updates Crackle extractor to allow downloading of entire series and to support a wider variety of URLs. Previously, a specific video ID had to be provided in the URL, but this is often not included in the URL when browsing the site normally. This change allows a channel or playlist ID to be provided in the URL instead.

URLs in the following formats are now supported:
 - ### Channel 
   - Format: https://www.crackle.com/watch/CHANNEL_ID
   - Channel can be a movie or a series. Extractor will check the JSON object for the channel to determine the type and download accordingly.
 - ### Playlist
   - Format: https://www.crackle.com/watch/playlist/PLAYLIST_ID
   - This will probably not be used much if at all in practice since Crackle tends to use channel IDs and/or specific video IDs in their links, but if a playlist URL is provided then the extractor will download all videos in the playlist.
 - ### Video
   - Formats include but are not limited to the following:
     - https://www.crackle.com/watch/playlist/PLAYLIST_ID/VIDEO_ID
     - https://www.crackle.com/watch/CHANNEL_ID/VIDEO_ID
   - Any channel or playlist ID will be ignored if a video ID is included and only the specific video will be downloaded.
   - Note: Some URLs which include the name of a series with a video ID are still supported but will redirect to a URL with a channel ID in it. I am guessing this was an older convention and they keep a mapping to handle routing to the new scheme.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
I would suggest splitting this into 3 extractors. The common code can be inherited from a `CrackleBaseIE`

**Comment by shloop** [comment]
> I would suggest splitting this into 3 extractors. The common code can be inherited from a `CrackleBaseIE`

Hmm, do you know of any other extractors that do that I could look to as an example? I was trying to mimic the logic flow of the PlutoTV one.

**Comment by shloop** [comment]
@pukkandan Refactored into separate extractors as requested.

File: yt_dlp/extractor/crackle.py [hunk 1]
```diff
@@ -13,16 +13,17 @@
...
-    _VALID_URL = r'(?:crackle:|https?://(?:(?:www|m)\.)?(?:sony)?crackle\.com/(?:playlist/\d+/|(?:[^/]+/)+))(?P<id>\d+)'
-    _TESTS = [{
-        # Crackle is available in the United States and territories
-        'url': 'https://www.crackle.com/thanksgiving/2510064',
-        'info_dict': {
+
+class CrackleBaseIE(InfoExtractor):
+
+    class TestData:
+        EPISODE_INFO = {
```
Comment (pukkandan) [inline yt_dlp/extractor/crackle.py]: Dont do this

File: yt_dlp/extractor/crackle.py [hunk 2]
```diff
@@ -39,19 +40,18 @@ class CrackleIE(InfoExtractor):
...
+            'season': 'Season 1',
+        }
+        EXPECTED_WARNINGS = ['Trying with a list of known countries']
+        PARAMS_SKIP_DOWNLOAD = {'skip_download': True}
+
+    class UrlType(Enum):
+        MEDIA = "media"
+        CHANNEL = "channel"
+        CHANNEL_PLAYLIST = "channel playlist"
+        PLAYLIST = "playlist"
```
Comment (pukkandan) [inline yt_dlp/extractor/crackle.py:29]: unnecessary. See my changes to similar abstraction below

**Comment by pukkandan** [comment]
My review on other PRs also apply here. Especially https://github.com/yt-dlp/yt-dlp/pull/4801#discussion_r962273141

File: yt_dlp/extractor/crackle.py [hunk 3]
```diff
@@ -242,3 +245,160 @@ def _real_extract(self, url):
...
+        'expected_warnings': ['Trying with a list of known countries']
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        playlist, country = self._download_crackle_details(CrackleBaseIE.UrlType.PLAYLIST, playlist_id)
+
+        videos = (self._get_video_info(media['MediaInfo']['Id'], country)
+                  for media in traverse_obj(playlist, ('Result', 'Medias', ...)))
+        return self.playlist_result(videos, playlist_id, traverse_obj(playlist, ('Result', 'Name')))
```
Comment (pukkandan) [inline yt_dlp/extractor/crackle.py:404]: Fix --flat-playlist by returning URLs here. Any metadata that requires additional network requests should not be returned here. core code will handle further extraction

File: yt_dlp/extractor/crackle.py [hunk 4]
```diff
@@ -242,3 +245,160 @@ def _real_extract(self, url):
...
+        channel, country = self._download_crackle_details(CrackleBaseIE.UrlType.CHANNEL, channel_id)
+        channel_type = channel['ChannelTypeId']
+
+        if channel_type == 1:  # Movies
+            return self._get_video_info(traverse_obj['FeaturedMedia']['ID'], country)
+
+        elif channel_type in (11, 13):  # TV Shows
+            playlist, country = self._download_crackle_details(CrackleBaseIE.UrlType.CHANNEL_PLAYLIST, channel_id, country)
+            videos = (
+                self._get_video_info(media['MediaInfo']['Id'], country)
```
Comment (pukkandan) [inline yt_dlp/extractor/crackle.py:378]: same

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by shloop** [comment]
Abandoning this since it was never accepted and Crackle website has changed to the point that neither the main branch nor my changes appear to work at all. Entire site may be DRM protected now: https://github.com/yt-dlp/yt-dlp/issues/5118

**Pull request closed without merging**